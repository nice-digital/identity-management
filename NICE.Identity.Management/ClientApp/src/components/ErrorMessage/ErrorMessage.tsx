import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps): React.ReactElement => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there&apos;s been an error!</p>
			</>
		)}

		{process.env.NODE_ENV === "development" && (
			<dl>
				<dt>Message:</dt>
				<dd>{props.error.message}</dd>
				<dt>Stack:</dt>
				<dd>{props.error.stack}</dd>
			</dl>
		)}
	</Alert>
);
