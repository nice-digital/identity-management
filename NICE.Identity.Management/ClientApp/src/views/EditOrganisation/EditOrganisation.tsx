import React, { Component } from "react";
import { RouteComponentProps, Link, Redirect } from "react-router-dom";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Input } from "@nice-digital/nds-input";
import { Endpoints } from "src/data/endpoints";
import { fetchData } from "src/helpers/fetchData";
import { isDataError } from "src/helpers/isDataError";
import { ErrorMessage } from "src/components/ErrorMessage/ErrorMessage";
import { type OrganisationType } from "src/models/types";

type TParams = { id: string };

type EditOrganisationProps = Record<string, unknown> &
	RouteComponentProps<TParams>;

type EditOrganisationState = {
	validationError: boolean;
	validationErrorMessage: string;
	fetchedOrgNameFound: boolean;
	organisation: OrganisationType;
	formName: string;
	isLoading: boolean;
	hasSubmitted: boolean;
	error?: Error;
	isSaveButtonLoading: boolean;
};
export class EditOrganisation extends Component<
	EditOrganisationProps,
	EditOrganisationState
> {
	constructor(props: EditOrganisationProps) {
		super(props);

		this.state = {
			validationError: false,
			validationErrorMessage:
				"Organisation name should be alphanumeric and be between 2-100 characters",
			fetchedOrgNameFound: false,
			organisation: {} as OrganisationType,
			formName: "",
			isLoading: true,
			hasSubmitted: false,
			isSaveButtonLoading: false,
		};

		document.title = "NICE Accounts - Edit organisation";
	}

	typingTimer = 0;

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const organisation = await fetchData(
			Endpoints.organisation(this.props.match.params.id),
		);

		if (isDataError(organisation)) {
			this.setState({ error: organisation });
		} else {
			this.setState({ formName: organisation.name });
		}

		this.setState({ organisation, isLoading: false });
	}

	checkOrgName = async (formName: string): Promise<void> => {
		const orgName = this.state.organisation.name.toLowerCase();
		formName = formName.toLowerCase();
		const namesAreDifferent = orgName !== formName;
		let fetchedOrgNameFound = false;

		if (namesAreDifferent) {
			const fetchOrgName = await fetchData(
				Endpoints.organisationsListSearch(formName),
			);

			if (isDataError(fetchOrgName)) {
				this.setState({ error: fetchOrgName });
			}

			fetchedOrgNameFound = fetchOrgName.length
				? fetchOrgName.some((org: any) => org.name.toLowerCase() === formName)
				: fetchedOrgNameFound;
		}

		this.setState({ fetchedOrgNameFound });
		this.typingTimer = 0;
	};

	toggleValidationMessage = (
		blockedOrgNameFound: boolean,
		orgName = "",
	): string => {
		const orgNameOrDefault = orgName || "organisation";
		return blockedOrgNameFound
			? `Cannot change to ${orgNameOrDefault}, that organisation already exists!`
			: "Organisation name should be alphanumeric and be between 2-100 characters";
	};

	handleSubmit = async (
		e: React.FormEvent<HTMLFormElement>,
	): Promise<void | boolean> => {
		e.preventDefault();
		this.setState({ isSaveButtonLoading: true });

		const { validationError } = this.state;
		const form = e.currentTarget;
		const formName = this.state.formName.trim();
		const formHasNotChanged = formName === this.state.organisation.name;

		if (formHasNotChanged) {
			this.setState({
				isSaveButtonLoading: false,
				hasSubmitted: true,
			});
			return false;
		}

		if (this.typingTimer) {
			clearTimeout(this.typingTimer);
			await this.checkOrgName(formName);
		}

		const { fetchedOrgNameFound } = this.state;

		if (!form.checkValidity() || validationError || fetchedOrgNameFound) {
			const validationErrorMessage = this.toggleValidationMessage(
				fetchedOrgNameFound,
				formName,
			);
			this.setState({
				validationError: true,
				validationErrorMessage,
				isSaveButtonLoading: false,
			});
			return false;
		}

		const fetchOptions = {
			method: "PATCH",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				name: formName,
			}),
		};

		const organisation = await fetchData(
			Endpoints.organisation(this.props.match.params.id),
			fetchOptions,
		);

		if (isDataError(organisation)) {
			this.setState({ error: organisation });
		}

		this.setState({
			isSaveButtonLoading: false,
			hasSubmitted: true,
		});
	};

	handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
		const formElement = e.target;
		const formName = formElement.value;
		const isNowValid = formElement.validity.valid;
		let { validationError } = this.state;

		if (isNowValid) {
			validationError = false;
		}

		clearTimeout(this.typingTimer);
		this.typingTimer = setTimeout(this.checkOrgName, 1000, formName.trim());

		this.setState({
			formName,
			validationError,
		});
	};

	handleBlur = async (
		e: React.ChangeEvent<HTMLInputElement>,
	): Promise<void> => {
		const formElement = e.target;
		const formName = this.state.formName.trim();

		if (this.typingTimer) {
			clearTimeout(this.typingTimer);
			await this.checkOrgName(formName);
		}

		const { fetchedOrgNameFound } = this.state;

		const isNowValid = formElement.validity.valid
			? !fetchedOrgNameFound
			: false;

		const validationErrorMessage = this.toggleValidationMessage(
			fetchedOrgNameFound,
			formName,
		);

		this.setState({
			validationError: !isNowValid,
			validationErrorMessage,
			formName,
		});
	};

	render(): JSX.Element {
		const {
			organisation,
			formName,
			isLoading,
			hasSubmitted,
			error,
			isSaveButtonLoading,
			validationError,
			validationErrorMessage,
		} = this.state;
		const organisationId = this.props.match.params.id;

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/overview" elementType={Link}>
						Administration
					</Breadcrumb>
					<Breadcrumb to="/organisations" elementType={Link}>
						Organisations
					</Breadcrumb>
					<Breadcrumb
						to={`/organisations/${organisationId}`}
						elementType={Link}
					>
						{isLoading ? "Loading organisation details" : organisation.name}
					</Breadcrumb>
					<Breadcrumb>Edit organisation</Breadcrumb>
				</Breadcrumbs>

				{error ? (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				) : (
					<>
						<PageHeader heading="Edit organisation" />
						{isLoading ? (
							<p>Loading...</p>
						) : (
							<>
								{hasSubmitted && (
									<Redirect
										to={{
											pathname: `/organisations/${organisationId}`,
											state: { hasBeenEdited: true },
										}}
									/>
								)}

								<Grid>
									<GridItem cols={12} md={6} lg={4}>
										<form onSubmit={this.handleSubmit} noValidate>
											<Input
												data-qa-sel="name-input-edit-organisation"
												label="Organisation name"
												name="orgName"
												type="text"
												required
												autoComplete="off"
												minLength="2"
												maxLength="100"
												pattern="^((?![<>]).)*$"
												onChange={this.handleChange}
												onBlur={this.handleBlur}
												error={validationError}
												errorMessage={validationErrorMessage}
												value={formName}
												disabled={isSaveButtonLoading}
											/>
											<Button
												data-qa-sel="save-button-edit-organisation"
												variant="cta"
												type="submit"
												disabled={isSaveButtonLoading}
											>
												{isSaveButtonLoading ? "Loading..." : "Save changes"}
											</Button>
											<Button
												to={`/organisations/${organisationId}`}
												variant="secondary"
												elementType={Link}
												disabled={isSaveButtonLoading}
											>
												Cancel
											</Button>
										</form>
									</GridItem>
								</Grid>
							</>
						)}
					</>
				)}
			</>
		);
	}
}
