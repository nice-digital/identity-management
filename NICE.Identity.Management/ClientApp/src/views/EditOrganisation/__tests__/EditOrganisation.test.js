import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import { Redirect } from "react-router-dom";
import toJson from "enzyme-to-json";
import { nextTick } from "../../../utils/nextTick";
import { Alert } from "@nice-digital/nds-alert";
import { EditOrganisation } from "../EditOrganisation";
import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";
import singleOrganisation from "./singleOrganisation.json";
import organisations from "./organisations.json";

describe("EditOrganisation", () => {
    const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

    it("should match the snapshot on load", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
        const wrapper = mount(<MemoryRouter><EditOrganisation match={match} /></MemoryRouter>);
        await nextTick();
        wrapper.update();
        expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should disable form submit button when clicked", async () => {
		console.error = jest.fn();
        fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify([]));
		const wrapper = mount(
			<MemoryRouter>
				<EditOrganisation match={match} />
			</MemoryRouter>,
		);
        await nextTick();
		wrapper.update();
		wrapper.find("#orgName").prop("onChange")({
			target: {
				name: "orgName",
				value: "Org Ninety Nine",
				validity: { valid: true }
			}
		});
		wrapper.find("form").simulate("submit");
		expect(wrapper.find("button").props().disabled).toEqual(true);
		expect(wrapper.find("button").text()).toEqual("Loading...");
	});

    it("should redirect if form is submitted without changes", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		const wrapper = mount(
			<MemoryRouter>
				<EditOrganisation match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").prop("onSubmit")({
			preventDefault: () => false, 
			currentTarget: { 
				checkValidity: () => true
			}
		});
		await nextTick();
		wrapper.update();
		expect(wrapper.find(Redirect).exists()).toBe(true);
	});

    it("should show error message when duplicate name fetchData get fails", async () => {
		console.error = jest.fn();
		const error = new Error("Not allowed");
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockRejectOnce(error);
		const wrapper = mount(
			<MemoryRouter>
				<EditOrganisation match={match} />
			</MemoryRouter>,
		);
        await nextTick();
		wrapper.update();
		wrapper.find("#orgName").prop("onChange")({
			target: {
				name: "orgName",
				value: "Org Ninety Nine",
				validity: { valid: true }
			}
		});
		wrapper.find("form").prop("onSubmit")({
			preventDefault: () => false, 
			currentTarget: { 
				checkValidity: () => true
			},
		});
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show validation error when name is invalid format", async () => {
		console.error = jest.fn();
        fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify([]));
		const wrapper = mount(
			<MemoryRouter>
				<EditOrganisation match={match} />
			</MemoryRouter>,
		);
        await nextTick();
		wrapper.update();		
		wrapper.find("#orgName").prop("onChange")({
			target: {
				name: "orgName",
				value: "a",
				validity: { valid: false }
			}
		});
		wrapper.find("#orgName").prop("onBlur")({
			target: {
				name: "orgName",
				value: "a",
				validity: { valid: false }
			}
		});
		await nextTick();
		wrapper.update();
		expect(wrapper.find({ label: "Organisation name" }).prop("error")).toEqual(true);
	});

	it("should show validation error when org name is in use already", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify(organisations));
		const wrapper = mount(
			<MemoryRouter>
				<EditOrganisation match={match} />
			</MemoryRouter>,
		);
        await nextTick();
		wrapper.update();
		wrapper.find("#orgName").prop("onChange")({
			target: {
				name: "orgName",
				value: "Org 2",
				validity: { valid: true }
			}
		});
		wrapper.find("#orgName").prop("onBlur")({
			target: {
				name: "orgName",
				value: "Org 2",
				validity: { valid: true }
			}
		});
		await nextTick();
		wrapper.update();
		expect(wrapper.find({ label: "Organisation name" }).prop("error")).toEqual(true);
		expect(wrapper.find({ label: "Organisation name" }).prop("errorMessage")).toBe("Cannot change to Org 2, that organisation already exists!");
	});

});