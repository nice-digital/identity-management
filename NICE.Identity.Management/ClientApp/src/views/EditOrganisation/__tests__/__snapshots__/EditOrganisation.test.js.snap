// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditOrganisation should match the snapshot on load 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/organisations"
            onClick={[Function]}
          >
            Organisations
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/organisations/1"
            onClick={[Function]}
          >
            Org 1
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Edit organisation
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"@id\\": \\"/organisations\\",
        \\"name\\": \\"Organisations\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 3,
      \\"item\\": {
        \\"@id\\": \\"/organisations/1\\",
        \\"name\\": \\"Org 1\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 4,
      \\"item\\": {
        \\"name\\": \\"Edit organisation\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Edit organisation
    </h1>
  </div>,
  <div
    className="grid"
  >
    <div
      data-g="12 md:6 lg:4"
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="input"
        >
          <label
            className="input__label"
            htmlFor="orgName"
          >
            Organisation name
          </label>
          <input
            autoComplete="off"
            className="input__input"
            data-qa-sel="name-input-edit-organisation"
            disabled={false}
            id="orgName"
            maxLength="100"
            minLength="2"
            name="orgName"
            onBlur={[Function]}
            onChange={[Function]}
            pattern="^((?![<>]).)*$"
            required={true}
            type="text"
            value="Org 1"
          />
        </div>
        <button
          className="btn btn--cta"
          data-qa-sel="save-button-edit-organisation"
          disabled={false}
          type="submit"
        >
          Save changes
        </button>
        <a
          className="btn btn--secondary"
          disabled={false}
          href="/organisations/1"
          onClick={[Function]}
        >
          Cancel
        </a>
      </form>
    </div>
  </div>,
]
`;
